src/cares_wrap.cc

```diff
@@ -1172,694 +580,622 @@ int ParseSoaReply(Environment* env,
+  const int timeout = args[0].As<Int32>()->Value();
+  Environment* env = Environment::GetCurrent(args);
+  new ChannelWrap(env, args.This(), timeout);
+}
 
-    CallOnComplete(ret, ttls);
-  }
-};
+GetAddrInfoReqWrap::GetAddrInfoReqWrap(
+    Environment* env,
+    Local<Object> req_wrap_obj,
+    bool verbatim)
+    : ReqWrap(env, req_wrap_obj, AsyncWrap::PROVIDER_GETADDRINFOREQWRAP),
+      verbatim_(verbatim) {}
 
+GetNameInfoReqWrap::GetNameInfoReqWrap(
+    Environment* env,
+    Local<Object> req_wrap_obj)
+    : ReqWrap(env, req_wrap_obj, AsyncWrap::PROVIDER_GETNAMEINFOREQWRAP) {}
 
-class QueryAaaaWrap: public QueryWrap {
- public:
-  QueryAaaaWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
-      : QueryWrap(channel, req_wrap_obj, "resolve6") {
-  }
+/* This is called once per second by loop->timer. It is used to constantly */
+/* call back into c-ares for possibly processing timeouts. */
+void ChannelWrap::AresTimeout(uv_timer_t* handle) {
+  ChannelWrap* channel = static_cast<ChannelWrap*>(handle->data);
+  CHECK_EQ(channel->timer_handle(), handle);
+  CHECK_EQ(false, channel->task_list()->empty());
+  ares_process_fd(channel->cares_channel(), ARES_SOCKET_BAD, ARES_SOCKET_BAD);
+}
 
-  int Send(const char* name) override {
-    AresQuery(name, ns_c_in, ns_t_aaaa);
-    return 0;
-  }
 
-  SET_NO_MEMORY_INFO()
-  SET_MEMORY_INFO_NAME(QueryAaaaWrap)
-  SET_SELF_SIZE(QueryAaaaWrap)
-
- protected:
-  void Parse(unsigned char* buf, int len) override {
-    HandleScope handle_scope(env()->isolate());
-    Context::Scope context_scope(env()->context());
-
-    ares_addr6ttl addrttls[256];
-    int naddrttls = arraysize(addrttls), status;
-    Local<Array> ret = Array::New(env()->isolate());
-
-    int type = ns_t_aaaa;
-    status = ParseGeneralReply(env(),
-                               buf,
-                               len,
-                               &type,
-                               ret,
-                               addrttls,
-                               &naddrttls);
-    if (status != ARES_SUCCESS) {
-      ParseError(status);
-      return;
-    }
+void NodeAresTask::MemoryInfo(MemoryTracker* tracker) const {
+  tracker->TrackField("channel", channel);
+}
 
-    Local<Array> ttls = AddrTTLToArray<ares_addr6ttl>(env(),
-                                                      addrttls,
-                                                      naddrttls);
+/* Allocates and returns a new NodeAresTask */
+NodeAresTask* NodeAresTask::Create(ChannelWrap* channel, ares_socket_t sock) {
+  auto task = new NodeAresTask();
 
-    CallOnComplete(ret, ttls);
-  }
-};
+  task->channel = channel;
+  task->sock = sock;
 
-class QueryCaaWrap: public QueryWrap {
- public:
-  QueryCaaWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
-      : QueryWrap(channel, req_wrap_obj, "resolveCaa") {
+  if (uv_poll_init_socket(channel->env()->event_loop(),
+                          &task->poll_watcher, sock) < 0) {
```

Comment:
Does this mean we are creating a new socket for every DNS resolution?

Reply:
No. We are creating a `uv_poll_t` here, not a socket.

Reply:
Even the one uv_poll_t per DNS resolution has it's own non-trivial cost. I'd prefer to find a more efficient solution but if @devsnek is successful in getting the transition to `getdns` going and landed, it's likely not worth the effort to refactor this.